diff -pruN signify-portable-20140902/Makefile signify-portable-20140902-h/Makefile
--- signify-portable-20140902/Makefile	2014-09-02 00:26:00.000000000 -0700
+++ signify-portable-20140902-h/Makefile	2014-10-25 03:02:49.628893485 -0700
@@ -8,6 +8,8 @@ mandir=${prefix}/man
 DESTDIR=
 
 CC=gcc
+AR=ar
+LIB=static
 INSTALL=/usr/bin/install -c
 
 CFLAGS=-O2 -D_FORTIFY_SOURCE=2 -ftrapv -fPIE -fstack-protector-all \
@@ -15,23 +17,28 @@ CFLAGS=-O2 -D_FORTIFY_SOURCE=2 -ftrapv -
 
 TARGET = signify
 
+TARGET_LIB = libsignify_$(LIB).a
+
 SIGNIFY_OBJS = signify.o fe25519.o sc25519.o smult_curve25519_ref.o \
 	mod_ed25519.o mod_ge25519.o crypto_api.o base64.o bcrypt_pbkdf.o \
 	explicit_bzero.o arc4random.o timingsafe_bcmp.o sha2.o blowfish.o \
 	readpassphrase.o strlcpy.o helpers.o ohash.o
 
-all:	$(TARGET)
+all:	$(TARGET) $(TARGET_LIB)
 
 $(TARGET): $(SIGNIFY_OBJS)
 	$(CC) $(CFLAGS) -o $(TARGET) $(SIGNIFY_OBJS)
 
-clean:	
-	rm -f *.o signify
+$(TARGET_LIB): $(SIGNIFY_OBJS)
+	$(AR) crs $(TARGET_LIB) $(SIGNIFY_OBJS)
+
+clean:
+	@rm -f *.o signify $(TARGET_LIB)
 
 install:
-	$(INSTALL) -D -m 0755 signify $(DESTDIR)/$(bindir)/signify
-	$(INSTALL) -D -m 644 signify.1 $(DESTDIR)/$(mandir)/man1/signify.1
+	$(INSTALL) -c -D -m 0755 signify $(DESTDIR)/$(bindir)/signify
+	$(INSTALL) -c -D -m 644 signify.1 $(DESTDIR)/$(mandir)/man1/signify.1
 
 uninstall:
-	rm -f $(DESTDIR)/$(bindir)/signify
-	rm -f $(DESTDIR)/$(mandir)/man1/signify.1
+	@rm -f $(DESTDIR)/$(bindir)/signify
+	@rm -f $(DESTDIR)/$(mandir)/man1/signify.1
